cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java

```diff
@@ -91,51 +201,730 @@ public void schedule(Id.Program program, SchedulableProgramType programType, Ite
+    }
   }
 
   @Override
   public void deleteSchedule(Id.Program programId, SchedulableProgramType programType, String scheduleName)
-    throws NotFoundException, SchedulerException {
-
+    throws ScheduleNotFoundException, SchedulerException {
+    String scheduleId = AbstractSchedulerService.scheduleIdFor(programId, programType, scheduleName);
+    StreamSubscriber subscriber = scheduleSubscribers.remove(scheduleId);
+    if (subscriber == null) {
+      throw new ScheduleNotFoundException(scheduleName);
+    }
+    subscriber.deleteSchedule(programId, programType, scheduleName);
+    // We don't delete a StreamSubscriber, even if it has zero task. We keep an empty subscriber so that we don't
+    // have to worry about race conditions between add/delete of schedules
   }
 
   @Override
-  public void deleteSchedules(Id.Program programId, SchedulableProgramType programType)
-    throws SchedulerException {
-
+  public void deleteSchedules(Id.Program programId, SchedulableProgramType programType) throws SchedulerException {
+    char startChar = ':';
+    char endChar = (char) (startChar + 1);
+    String programScheduleId = AbstractSchedulerService.programIdFor(programId, programType);
+    NavigableSet<String> scheduleIds = scheduleSubscribers.subMap(String.format("%s%c", programScheduleId, startChar),
+                                                                  String.format("%s%c", programScheduleId, endChar))
+      .keySet();
+    int scheduleIdIdx = programScheduleId.length() + 1;
+    for (String scheduleId : scheduleIds) {
+      try {
+        if (scheduleId.length() < scheduleIdIdx) {
+          LOG.warn("Format of scheduleID incorrect: {}", scheduleId);
+          continue;
+        }
+        deleteSchedule(programId, programType, scheduleId.substring(scheduleIdIdx));
+      } catch (ScheduleNotFoundException e) {
+        // Could be a race, the schedule has just been deleted
+        LOG.debug("Could not delete schedule, it might have been deleted already by another thread '{}'",
+                  scheduleId, e);
+      }
+    }
   }
 
   @Override
   public ScheduleState scheduleState(Id.Program program, SchedulableProgramType programType, String scheduleName)
     throws SchedulerException {
-    return ScheduleState.NOT_FOUND;
+    StreamSubscriber subscriber = scheduleSubscribers.get(AbstractSchedulerService.scheduleIdFor(program, programType,
+                                                                                                 scheduleName));
+    if (subscriber != null) {
+      return subscriber.scheduleTaskState(program, programType, scheduleName);
+    } else {
+      return ScheduleState.NOT_FOUND;
+    }
   }
 
-  private String getScheduleId(Id.Program program, SchedulableProgramType programType, String scheduleName) {
-    return String.format("%s:%s", getProgramScheduleId(program, programType), scheduleName);
+  private synchronized Store getStore() {
```

Comment:
doesn't it make more sense to create this once in the constructor? Since this store will be used all the time, I don't see a lot of value in lazy instantiation.